src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1205 @@
+			entry->ae_par_extent.ape_epoch != ~(0ULL)) {
+		/* Parity newer than data; nothing to do. */
+		goto out;
+	}
+
+	if ((entry->ae_par_extent.ape_epoch == ~(0ULL)
+				 && agg_stripe_is_filled(entry, false)) ||
+					agg_stripe_is_filled(entry, true)) {
+		/* Replicas constitute a full stripe. */
+		rc = agg_encode_local_parity(entry);
+		goto out;
+	if (entry->ae_par_extent.ape_epoch == ~(0ULL))
+		update_vos = false;
+		goto out;
+	}
+	/* Parity, some later replicas, not full stripe. */
+	rc = agg_process_partial_stripe(entry);
+out:
+	if (update_vos)
+		rc = agg_update_vos(entry);
+		/* offload of ds_obj_update (TBD) to push remote parity */
+
+	agg_clear_extents(entry);
+	return rc;
+}
+
+/* returns the subrange of the RECX iterator's returned recx that lies within
+ * the current stripe.
+ */
+static daos_off_t
+agg_in_stripe(struct ec_agg_entry *entry, daos_recx_t *recx)
+{
+	unsigned int		len = entry->ae_oca->u.ec.e_len;
+	unsigned int		k = entry->ae_oca->u.ec.e_k;
+	daos_off_t		stripe = recx->rx_idx / (len * k);
+	daos_off_t		stripe_end = (stripe + 1) * len * k;
+
+	if (recx->rx_idx + recx->rx_nr > stripe_end)
+		return stripe_end - recx->rx_idx;
+	else
+		return recx->rx_nr;
+}
+
+/* Iterator call back sub-function for handling data extents.
+ */
+static int
+agg_data_extent(vos_iter_entry_t *entry, struct ec_agg_entry *agg_entry,
+		daos_handle_t ih, unsigned int *acts)
+{
+	struct ec_agg_extent	*extent = NULL;
+	int			 rc = 0;
+
+	if (agg_stripenum(agg_entry, entry->ie_recx.rx_idx) !=
+			agg_entry->ae_cur_stripe.as_stripenum) {
+		if (agg_entry->ae_cur_stripe.as_stripenum != ~0UL)
+			agg_process_stripe(agg_entry);
+		agg_entry->ae_cur_stripe.as_stripenum =
+			agg_stripenum(agg_entry, entry->ie_recx.rx_idx);
+	}
+	D_ALLOC_PTR(extent);
+	if (extent == NULL) {
+		rc = -DER_NOMEM;
+		goto out;
+	}
+	extent->ae_recx = entry->ie_recx;
+	extent->ae_epoch = entry->ie_epoch;
+	agg_entry->ae_rsize = entry->ie_rsize;
+
+	d_list_add_tail(&extent->ae_link,
+			&agg_entry->ae_cur_stripe.as_dextents);
+	agg_entry->ae_cur_stripe.as_extent_cnt++;
+	agg_entry->ae_cur_stripe.as_stripe_fill +=
+		agg_in_stripe(agg_entry, &entry->ie_recx);
+
+	if (extent->ae_epoch > agg_entry->ae_cur_stripe.as_hi_epoch)
+		agg_entry->ae_cur_stripe.as_hi_epoch = extent->ae_epoch;
+
+	D_PRINT("adding extent %lu,%lu, to stripe  %lu, shard: %u\n",
+		extent->ae_recx.rx_idx, extent->ae_recx.rx_nr,
+		agg_stripenum(agg_entry, extent->ae_recx.rx_idx),
+		agg_entry->ae_oid.id_shard);
+out:
+	return rc;
+}
+
+static int
+agg_akey_post(daos_handle_t ih, vos_iter_entry_t *entry,
+	 struct ec_agg_entry *agg_entry,
+	 unsigned int *acts)
+{
+	int rc = 0;
+
+	if (agg_entry->ae_cur_stripe.as_extent_cnt)
+		agg_process_stripe(agg_entry);
+
+	return rc;
+}
+
+/* Handles each replica extent returned by the RECX iterator.
+ */
+static int
+agg_ev(daos_handle_t ih, vos_iter_entry_t *entry,
+	struct ec_agg_entry *agg_entry, unsigned int *acts)
+{
+	int			rc = 0;
+
```

Comment:
D_ASSERT((entry->ie_recx.rx_idx & PARITY_INDICATOR) == 0);

Then you can get rid of if/else

Reply:
This is a holdover from before I had the range filter on the iterator.  I will change it.